chronos/asyncfutures2.nim

```diff
@@ -317,6 +317,9 @@ proc futureContinue*(fut: FutureBase) {.raises: [], gcsafe.} =
       # `await` typically) or completes / fails / is cancelled
       next = fut.internalClosure(fut)
       if fut.internalClosure.finished(): # Reached the end of the transformed proc
+        if not fut.finished:
```

Comment:
hm, this is actually a really good point: we should not finish the future until the iterator itself has finished or an order inversion will indeed happen! it deserves to be documented.

it means that the above should not be a conditional but rather the norm - we should always arrive at this point with an unfinished future.

That in turn means that the expression return stuff needs to be rewritten as well to assign to value rather than calling `complete` - ditto error paths which need to set the error field but not complete the future etc..

Reply:
it's actually a classic bug not to run iterators until completion and thus missing part of the code - the same happens with simple `for` loop iterators that have code after the last `yield`: if you break the for loop that code will not run..

Reply:
> setFinishLocation

or just use `internalLocation[...]` - ie none of this should be exposed in any public functions or API so it doesn't greatly matter - we can design a better public API for these kinds of things once the dust has settled

Reply:
Done

Reply:
Note that this slightly changed the Finish location (the `proc` line instead of the first line of the proc)
Not sure this greatly matters